docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+  avoiding sharing cross-cluster secrets in a single secrets blob
+
+#### Pros alternative-2
+
+- Retains all of the prior pros in Alternative-1
+- Does not leak secrets across separate Ceph clusters, possibly managed by
+  different namespaces as there exists no *ceph-cluster-secrets*
+
+#### Cons alternative-2
+
+- *ceph-cluster-secrets* is absent and hence following operations **cannot**
+  be supported
+  - ListVolumes, ListSnapshots, and GetCapacity
+  - **NOTE:** Unless listing rbd and CephFS volumes can be done without
+    credentials
+- Bloats every StorageClass making secrets mandatory
+- Retains other cons as in Alternative-1
+
+### Solution Alternative-3
+
+#### Summary alternative-3
+
+- Address *ceph-clusters* being accessible to all namespaces that need to
+  manage it
+- Move the information in *ceph-clusters* to secrets in the StorageClass (as in
+  monValueFromSecret configuration implementation in the code as of this
+  writing)
+
+#### Pros and cons alternative-3
+
+- Cons remains the same, with further added complexity in the StorageClass and
+  its secrets
+- Removes the con in Alternate-1 and 2 where the *ceph-clusters* information is
+  accessible across namespaces
+
+### Solution Alternative-4
+
+#### Summary alternative-4
+
+- Create a CSI plugin instance per Ceph cluster, with its own unique provisioner
+  name
+- Each CSI plugin instance can now have its own *ceph-cluster* and
+  *ceph-secrets* config maps, where *ceph-cluster* contains information
+  regarding exactly one cluster and similarly the *ceph-secret* for the same
+- StorageClass points to the provisioner name and does not need secrets,
+  cluster-id and may optionally choose to contain pool-name (can be omitted,
+  if we choose to to allocate from an available pool in the cluster rather
+  than a specific pool, based on the CreateVolume request parameters and
+  Topology)
+
+#### Pros alternative-4
+
+- Can support all Operations, as cluster information and secrets are per
+  instance
+- The secrets and cluster information need not be shared across multiple
+  namespaces
+- StorageClass becomes leaner, but still needs at least as many StorageClasses
+  as there are clusters backing the CO environment
+- Allows for future merging of the CSI instances, reverting to the StorageClass
+  as in the proposed solution
+
+#### Cons alternative-4
+
+- We will have as many NodeService entities as there are clusters in each node
+  of the CO environment. This may consume more resources
+- We cannot, deal with multiple Ceph clusters backed by a single CSI plugin
+  instance, due to the strict 1:1 mapping between the same, and thus do any
+  form of intelligent provisioning across Ceph clusters for a request
+- **Needs Investigation** Does this hinder Topology based volume provisioning?
+  As each instance is stuck to single Ceph cluster, we cannot provision from
+  different Topologies (each with their own Ceph clusters)
```

Comment:
What is a Topology and how does it relate to a StorageClass?  Each StorageClass maps to a particular pool or fs in a particular ceph cluster, right?

Reply:
This surely needs more investigation from my end, so answers maybe sketchy.

First up, IMO even the CSI k8s implementation is sketchy in this regard (it is alpha). ([reference](https://kubernetes-csi.github.io/docs/topology.html)).

Interesting point to note is point (2) in the above link that reads,
```
Ability for Kubernetes (users or components) to influence where a volume is
provisioned (e.g. provision new volume in either "zone 1" or "zone 2").
```

I read the above as, a request may come in with a Topology requirement. If the StorageClass already has the cluster and pool specified, the StorageClass is tied to a Topology in my understanding and if so we cannot decide where to provision from, it is decided by the StorageClass already. Which means we lose the ability to provision from a specific Topology, and each pod has to use a StorageClass that ties it to a Topology.

Unless, as @JohnStrunk asks, and I have questions as well around the same, can a Ceph cluster span so called Typologies? IOW, let's call them data centers (DC) in a federated cluster and can we have pools on DC1 and different pools in DC2, but all part of the same Ceph cluster?
- If yes, then specifying the cluster in the StorageClass would do, the request would come in with a Topology requirement, and pool selection should be done based on the same
- If no, then CSI driver would have to choose the cluster itself to provision from, and hence even that information should not come from the StorageClass

All this is speculation at best, currently the StorageClass **will** specify the cluster and pool that the volume should be provisioned from. In the future, based on the above proposal and based on better understanding of what Topology really entails, we can omit the same and add the logic within the CSI driver if needed.